Legacy code removal in pass manager

nkanazawa1989 opened the pull request:
### Summary

This PR is a followup of #10127 in which several methods and classes were deprecated. This PR removes deprecated code, and introduces new pattern to build pass pipeline.

### Details and comments

The main goal of this PR is to remove following pattern to a build pass pipeline.

```python
pm = PassManager()
pm.append([my_pass1, my_pass2, my_pass3], condition=callable1, do_while=callable2)
```

This pattern implicitly calls a factory function to create flow controller instances and wrap the passes `my_pass*` with them, by digesting the controller namespace defined in `FlowController` class attributes. This also internally tracks the group of passes appended together along with the specified controller names, with the extra overhead of memory footprint., i.e. `PassManager._pass_sets`.

With this PR, this append pattern is almost eliminated. Users now need to explicitly instantiate the controller class instead of using the syntactic sugar above. This drastically simplifies the factory logic, and the user can also precisely manage the building of nested controllers.

In addition to this, following classes and methods are also removed.
- .append method of flow controller subclasses 937567ad3a0dc8d3347e3b791bdde0419275a031
- RunningPassManager which is a subclass of FlowControllerLinear e02f6f50b276374e0f0e5538f77afa71ffc53f06
- Fenced objects a6c491619251156f892e07f0f854b664ac6283ae
- max_iteration argument in append and replace 3b1161eacef39280ca907daec5e8db0c2f83c306

nkanazawa1989 committed:
Remove flow controller subclass .append

nkanazawa1989 committed:
Update builtin pass managers to avoid append pattern

nkanazawa1989 committed:
Remove running pass manager

nkanazawa1989 committed:
Remove PropertySet and Fenced objects from transpile

nkanazawa1989 committed:
Remove _pass_sets and let PassManager generate pass list on the fly

This is necessary because kwargs is going to be removed from append method.
Current pass list is generated based on the kwargs.

nkanazawa1989 committed:
Remove kwargs from PassManager.append and .replace

This also removes FlowController factory class

nkanazawa1989 committed:
Remove max_iteration keyword from append and replace

nkanazawa1989 committed:
Reno

nkanazawa1989 committed:
minor: fix test and docs

jakelishman commented on qiskit/passmanager/flow_controllers.py:
```diff
@@ -161,176 +116,10 @@ def passes(self) -> list[Task]:
-            raise KeyError("Flow controller not found: %s" % name)
-        del cls.registered_controllers[name]
-        cls.hierarchy.remove(name)
+FlowController = BaseController
```
Do we need to keep this name around now, or can we just rely on `BaseController`?

nkanazawa1989 replied:
Not necessary. The alias is removed in [6e6d90e](https://github.com/Qiskit/qiskit/pull/11448/commits/6e6d90e772efa2b879e3924a9ebc14fe5479d083). I was just worried about deprecation of import path.

jakelishman commented on qiskit/transpiler/passmanager.py:
```diff
@@ -579,39 +447,95 @@ def _wrapped_callable(task, passmanager_ir, property_set, running_time, count):
+      pm = PassManager()
+      pm.append([PassA, PassB])
+      pm.append([PassC, PassD], condition=callable)
+      pm.append([PassE])
```
If this function is purely for visualisation, it should be in the `pass_manager_visualization` module, most likely.  We should probably deprecate and remove `PassManager.passes` in favour of the flat `PassManager.tasks`, which I think already exists?

We shouldn't attempt to keep the recursive list structure around into the future; we already know that it's insufficient to represent even existing flow controllers, let alone future ones in the new structure.

nkanazawa1989 replied:
Yeah this makes perfect sense. Actually drawer doesn't need to rely on this pass representation. This method and representation are removed in [e3a991b](https://github.com/Qiskit/qiskit/pull/11448/commits/e3a991ba85e8c56cb027f4e72e9b81397da0e94f). I also need to update #11449 to add deprecation for the method.

jakelishman reviewed (commented):
For the most part this looks sensible to me, though it's rather hard to verify that the preset passmanagers have stayed logically the same; it perhaps would have been easier to do the simple one-to-one replacements in this PR, and have a follow-up that does the refactoring of the logic, so the two can be checked separately.  I tried to check them, but I'm not certain I'd have spotted a potential mistake.

Other than that, mostly just the additional question about `transpiler.PassManager.passes` - I'm not certain that keeping around the list format is a good idea, especially if it's in principal only for the visualisation.  Exposing a `tasks()` method that returns a copy of the raw underlying list seems fine, though (maybe even on `BasePassManager`).

nkanazawa1989 committed:
Remove qiskit.passmanager.flow_controllers.FlowController

nkanazawa1989 committed:
Remove .passes method from PassManager and StagedPassManager

nkanazawa1989 committed:
Update reference image of pass manager drawer test

nkanazawa1989 committed:
minor: fix drawer label generation

nkanazawa1989 committed:
Revert "Update builtin pass managers to avoid append pattern"

This reverts commit 4bcfcd68ae9fc5b8a9bfbdfad22aaa663ac972f3.

nkanazawa1989 committed:
Minimum update of builtin pass managers

nkanazawa1989 commented:
I reverted the commit for builtin passmanagers in [98c2ccf](https://github.com/Qiskit/qiskit/pull/11448/commits/98c2ccfc6675a94b28786a0c06934fbe9d6f3101) and added minimum change to them just to avoid deprecated code in [2f48717](https://github.com/Qiskit/qiskit/pull/11448/commits/2f48717995dbd6fca569f8a5a779e701327e8c9d). In the followup PR #11494 builtin pass managers are refactored to eliminate use of the append pattern. This PR must be merged first, then I'll rebase the followup PR.

jakelishman commented on releasenotes/notes/upgrade-pass-manager-98aa64edde67b5bb.yaml:
```diff
@@ -0,0 +1,79 @@
+        )
+      )
+
+    Note that you can manage the pecking order of controllers when you want to nest them,
```
> pecking order

I love this description!

jakelishman commented on test/python/transpiler/test_pass_scheduler.py:
```diff
@@ -589,21 +615,6 @@ def test_nested_conditional_in_loop(self):
-            for pass_ in self.passes:
-                metadata = yield pass_
-
-
```
This is removed because it's tested as part of `qiskit.passmanager` elsewhere, I assume?

jakelishman reviewed (approved):
Thanks Naoki for making the modifications to this - it was 100x easier to check the changes to the preset pass managers, and it's clear that the vast majority of cases have 1:1 replacements that are strictly more powerful now.
